nixos/roundcube: fix remote postgresql server

everlanes opened the pull request:
###### Description of changes
Roundcube did not start with remote PostgreSQL server.

Fixed two bugs:

  1. Password file for database access is used in two places which needed differing formats (#215986).
  2. The roundcube-setup service was startet before network was available and therefore failing.
    Service is run after network-online.target now.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Built an ISO image for KVM guest with `nixos-generate`
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

everlanes committed:
nixos/roundcube: fix PostgreSQL password

Extract PostgreSQL database password for Roundcube from .pgpass file.
The password file is used in two locations:

  1. in the Roundcube config.php
  2. in the systemd setup service that initializes the roundcube
     database

These two services need the password in different formats.

Keep the password file in PostgreSQL standard format and extract the
password for the Roundcube config (see #215986).

everlanes committed:
nixos/roundcube: fix roundcube-setup start

When using Roundcube with a non local PostgreSQL database wait for
network start before running roundcube-setup.service
Otherwise the database is not reachable and the service fails.

everlanes commented:
@symphorien I created a pull request for the issue #215986

Extracting the password should honour escaped characters.
Also fixed an issue with service startup sequence.

symphorien commented on nixos/modules/services/mail/roundcube.nix:
```diff
@@ -123,7 +123,15 @@ in
 
-      ${lib.optionalString (!localDB) "$password = file_get_contents('${cfg.database.passwordFile}');"}
+      ${lib.optionalString (!localDB) ''
+        # Password file should be formated according to PostgreSQL .pgpass standard
+        # see https://www.postgresql.org/docs/current/libpq-pgpass.html
```
this comment should go in the description of the option.

symphorien reviewed (commented):
how did you test? I don't have the correct setup to test these changes effectively.

symphorien commented:
@ofborg test roundcube

everlanes commented:
> how did you test? I don't have the correct setup to test these changes effectively.

I have a KVM host with two guest - basically the setup I wanted to create: one with PostgreSQL and one with Roundcube.

everlanes committed:
nixos/roundcube: extend documentation for passwordFile

The passwordFile must be formatted correctly. Added some information
about how the file content should look like.

symphorien commented:
Rereading the documentation of .pgpass, the file may contain several lines and commented lines. We can document that several lines are not supported, but this PR should probably be amended to strip newlines from passwords (many text editors silently add trailing newlines to files) and to ignore comments.

everlanes commented:
The password file in my test environment has a trailing newline and it is working fine.

Parsing the password file like PostgreSQL would be nice, but also lead into a rabbit hole I don't want to enter: matching host, port and user against wildcards and config values, escaping the config values correctly, ignoring comment lines, etc. This sound quite error prone and not worth the effort.

I would suggest, to add a warning to the documentation, that only a single line and no comments are supportet. I could also replace `file_get_contents` with `file(..)[0]` to only read the first line.

symphorien commented:
> I would suggest, to add a warning to the documentation, that only a single line and no comments are supportet. I could also replace file_get_contents with file(..)[0] to only read the first line.

Either is fine by me.

everlanes committed:
nixos/roundcube: read only first line of password file

Only read the first line of the password file, if has multiple lines.
Mention that behaviour in options documentation.

symphorien commented:
Thank you!
